### PR TITLE
(PCP-835) Add support for task implementations

### PR DESCRIPTION
**Pull request by MikaelSmith**
Select implementation that matches known features, initially only `puppet-agent`. Uses first file in `files` if no implementations are provided. Errors if implementations are provided but features do not satisfy any of their requirements.

### CODE REVIEW

**Comment by MikaelSmith** [comment]
I'll plan to add an acceptance test around making sure we download the right files.

**Comment by MikaelSmith** [comment]
jenkins, please test this

**Comment by MikaelSmith** [comment]
jenkins please test this

**Comment by MikaelSmith** [comment]
jenkins please test this

**Comment by MikaelSmith** [comment]
jenkins please test this on aix71-POWER,osx1012-64,redhat7-64,sles12-64,solaris10-32,solaris11-64,windows2012r2-64

**Comment by MikaelSmith** [comment]
jenkins please test this on aix61-POWER,osx1012-64,redhat7-64,sles12-64,solaris10-32,solaris11-64,windows2012r2-64

**Comment by MikaelSmith** [comment]
jenkins please test this on redhat7-64,windows2012r2-64

**Comment by MikaelSmith** [comment]
I can't reproduce the run_ps1 failure, I don't think it's caused by my changes. Encountering errors starting the service with Puppet.

**Comment by MikaelSmith** [comment]
jenkins please test this on redhat7-64a,windows2012r2-64a

**Comment by MikaelSmith** [comment]
So turns out the `test on` logic is very literal. If we don't include agents, must of the tests do nothing. The run_ps1 test was poorly written, so it tries to run pxp-agent on non-agent Windows targets. That works generally because Windows is never a master platform, but fails if the windows hosts aren't agents.

**Comment by MikaelSmith** [comment]
Acceptance tests passed.

File: acceptance/tests/tasks/task_implementations.rb [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    # Make the static content path
+    env_name = File.basename(__FILE__, '.*')
+    @static_content_path = File.join(environmentpath, mk_tmp_environment(env_name))
+
+    # Create the task
+    task_body = "#!/opt/puppetlabs/puppet/bin/ruby\nputs ENV['PT_message']"
+    taskpath = "#{@static_content_path}/#{filename}"
+    create_remote_file(master, taskpath, task_body)
+    on master, "chmod 1777 #{taskpath}"
+    @sha256 = Digest::SHA256.hexdigest(task_body + "\n")
```
Comment (nicklewis) [inline acceptance/tests/tasks/task_implementations.rb]: Where does the `\n` come from in the task body?

Comment (MikaelSmith) [reply acceptance/tests/tasks/task_implementations.rb]: I think that's automatically created on the target with `create_remote_file`. This is just copied around at this point.

Comment (nicklewis) [reply acceptance/tests/tasks/task_implementations.rb]: :+1:

**Pull request merged**